### PR TITLE
Kballsearcher

### PR DESCRIPTION
**Pull request by exepulveda**
Implementing KBallSearcher as we discussed. 
KBallSearcher uses knn and radius combined by searching for knn first and then filtering the resutls to locations withtin the radius.

### CODE REVIEW

File: src/neighborsearch/kball_searcher.jl [hunk 1]
```diff
@@ -0,0 +1,52 @@
+# ------------------------------------------------------------------
+# Licensed under the ISC License. See LICENSE in the project root.
+# ------------------------------------------------------------------
+
+"""
+KBallSearcher(object, neighborhood)
```
Comment (juliohm) [inline src/neighborsearch/kball_searcher.jl:6]: Could you please prepend 4 whitespaces to this line? This way the Julia help prompt will highlight the function signature.

File: src/neighborsearch/kball_searcher.jl [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+  searcher::S
+  nmax::Int
+end
+
+object(searcher::KBallSearcher) = object(searcher.searcher)
+
+maxneighbors(searcher::KBallSearcher) = searcher.nmax
+
+function search!(neighbors::AbstractVector{Int}, xₒ::AbstractVector,
+  searcher::KBallSearcher; mask=nothing)
```
Comment (juliohm) [inline src/neighborsearch/kball_searcher.jl:23]: [Pedantic mode activated]
Could you please align this line to fit below the first argument (in this case the "n" of "neighbors")
[Pedantic mode deactivated]

File: src/neighborsearch/kball_searcher.jl [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+"""
+KBallSearcher(object, neighborhood)
+
+A method for searching neighbors in spatial `object` inside `neighborhood`.
+Used when doing estimation combining k-nearest and radius.
+It searches first for the k-nearest by using the internal kdtree and then
+limiting the results within the `BallNeighborhood` by calling `isneighbor`
+"""
+struct KBallSearcher{S<:AbstractNeighborSearcher} <: AbstractBoundedNeighborSearcher
+  searcher::S
```
Comment (juliohm) [inline src/neighborsearch/kball_searcher.jl:14]: Should we make use of a more restrictive `S <: NearestNeighborSearcher`? I think the idea of this searcher only works with nearest-neighbor search as the inner searcher right?

File: src/neighborsearch/kball_searcher.jl [hunk 4]
```diff
@@ -0,0 +1,52 @@
...
+function search!(neighbors::AbstractVector{Int}, xₒ::AbstractVector,
+  searcher::KBallSearcher; mask=nothing)
+
+  knn_searcher = searcher.searcher #internar nearest searcher
+  k = searcher.nmax
+  neigh = knn_searcher.neigh       #the Ball Neighborhood
+  object = knn_searcher.object
+  radius = neigh.radius
+
+  inds, dists = knn(knn_searcher.kdtree, xₒ, k, true)
```
Comment (juliohm) [inline src/neighborsearch/kball_searcher.jl:31]: I think what you really want to do here is leverage the inner searcher and call the `search!` method for it?

```julia
locs = search(xₒ, searcher.searcher, mask=mask)
```
And then post-process the locations with `isneighbor`?

File: src/neighborsearch/kball_searcher.jl [hunk 5]
```diff
@@ -0,0 +1,52 @@
...
+Used when doing estimation combining k-nearest and radius.
+It searches first for the k-nearest by using the internal kdtree and then
+limiting the results within the `BallNeighborhood` by calling `isneighbor`
+"""
+struct KBallSearcher{S<:AbstractNeighborSearcher} <: AbstractBoundedNeighborSearcher
+  searcher::S
+  nmax::Int
+end
+
+object(searcher::KBallSearcher) = object(searcher.searcher)
```
Comment (juliohm) [inline src/neighborsearch/kball_searcher.jl:18]: I think this isn't needed if you just call the `search` method of the inner `searcher` directly. See comment bellow.

File: src/neighborsearch/kball_searcher.jl [hunk 6]
```diff
@@ -0,0 +1,52 @@
...
+KBallSearcher(object, neighborhood)
+
+A method for searching neighbors in spatial `object` inside `neighborhood`.
+Used when doing estimation combining k-nearest and radius.
+It searches first for the k-nearest by using the internal kdtree and then
+limiting the results within the `BallNeighborhood` by calling `isneighbor`
+"""
+struct KBallSearcher{S<:AbstractNeighborSearcher} <: AbstractBoundedNeighborSearcher
+  searcher::S
+  nmax::Int
```
Comment (juliohm) [inline src/neighborsearch/kball_searcher.jl:15]: Maybe the second field of the KBallSearcher is the ball? So you first search with the inner searcher, and then post-process the results with the ball, makes sense? No need to specify `nmax` because that is already inside the inner nearest-neighbor searcher.

File: src/neighborsearch/kball_searcher.jl [hunk 7]
```diff
@@ -0,0 +1,52 @@
...
+    if masked 
+      coordinates!(x, object, loc)
+      if isneighbor(neigh, xₒ, x) #limiting to the radius
+        #this may be improve as the knn already returns the distances, but less generic 
+        #because the BallNeighborhood can use a different metric indeed
+        nneigh += 1
+        neighbors[nneigh] = loc
+      end
+    end
+  end
```
Comment (juliohm) [inline src/neighborsearch/kball_searcher.jl:49]: This post-processing can be simplified after the other changes above take place.

File: test/neighborsearch.jl [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+    # KBallSearcher
+
+    #create a regular grid to test
+    pdomain = RegularGrid(10,10)
+
+    #test 1: all points should be found
+    maxneighbors = 10
+    neigh = BallNeighborhood(100.0)
+    searcher  = NeighborhoodSearcher(pdomain, neigh)
+    kballsearcher = KBallSearcher(searcher,maxneighbors)
```
Comment (juliohm) [inline test/neighborsearch.jl:12]: So the example I had in mind when you explained the KBallSearcher concept was something like:

```julia
knnsearcher = NearestNeighborSearcher(pdomain, maxneighbors)
kballsearcher = KBallSearcher(knnsearcher, neigh)
```
So the inner searcher is always a `NearestNeighborSearcher` and the other missing step is the `neigh` (ball). Makes sense?

File: test/neighborsearch.jl [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+    xₒ = [5.0,5.0]
+    nneigh = search!(neighbors,xₒ,kballsearcher; mask=nothing)
+
+    @test nneigh == 5
+
+    #test 3: radius of 1.0 m but far just to find 1 point
+    maxneighbors = 10
+    neigh = BallNeighborhood(1.0)
+    searcher  = NeighborhoodSearcher(pdomain, neigh)
+    kballsearcher = KBallSearcher(searcher,maxneighbors)
```
Comment (juliohm) [inline test/neighborsearch.jl:36]: Previous comment applies here too.

File: test/neighborsearch.jl [hunk 10]
```diff
@@ -0,0 +1,57 @@
...
+    xₒ = [-0.2,-0.2]
+    nneigh = search!(neighbors,xₒ,kballsearcher; mask=nothing)
+
+    @test nneigh == 1
+
+    #test 4: radius of 1.0 m but far just to find no points
+    maxneighbors = 10
+    neigh = BallNeighborhood(1.0)
+    searcher  = NeighborhoodSearcher(pdomain, neigh)
+    kballsearcher = KBallSearcher(searcher,maxneighbors)
```
Comment (juliohm) [inline test/neighborsearch.jl:48]: And here as well.

**Comment by juliohm** [comment]
Just in case you are not 100% familiar with the GitHub workflow, if you push new commits to the same branch on your fork, the PR will be automatically updated.

File: src/neighborsearch/kball_searcher.jl (hunk 4 above)
Comment (exepulveda) [reply src/neighborsearch/kball_searcher.jl:31]: We could do that, but in that way we would need to recalculate distances (search function does not return distances that has already done by the inner kdtree). That is the reason for implementing this as it is.

File: src/neighborsearch/kball_searcher.jl (hunk 6 above)
Comment (exepulveda) [reply src/neighborsearch/kball_searcher.jl:15]: That is one option. I just tried to reuse as much as possible what is done. For user perspective, it could be easier to just indicate something like this:
solver = Kriging(
    :cu => (variogram=vmodel,maxneighbors=25,radius=50.0)
)
but I guess we don't want to change the API.

File: src/neighborsearch/kball_searcher.jl (hunk 4 above)
Comment (juliohm) [reply src/neighborsearch/kball_searcher.jl:31]: What we can do then is consider `KBallSearcher` its own implementation without relying on other existing searchers. In this case, it could have two fields `ball` and `k`. The constructor instantiates the kdtree similar to what is done in the NearestNeighborSearcher constructor, and then we implement the search explicitly without an inner searcher. What do you think?

File: src/neighborsearch/kball_searcher.jl (hunk 6 above)
Comment (juliohm) [reply src/neighborsearch/kball_searcher.jl:15]: Yes, I think it is more general to leave the option as neighborhood because then we have other kinds of neighborhood that could be useful that are not norm-balls.

File: src/neighborsearch/kball_searcher.jl (hunk 4 above)
Comment (exepulveda) [reply src/neighborsearch/kball_searcher.jl:31]: Much better I think. For example:
` Kriging(
:cu => (variogram=vmodel,searcher=KBallSearcher(k=25,radius=50.0))
)`

Comment (juliohm) [reply src/neighborsearch/kball_searcher.jl:31]: I think the Kriging interface is another issue that we could consider later. Let's first work on this PR to get the KBallSearcher in place, then we proceed to the other PR to figure out an optimal set of parameters 👍

File: src/neighborsearch/kball_searcher.jl [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+
+
+function KBallSearcher(object::O, k::Int, ball::BallNeighborhood{T,M}; locations=1:npoints(object)) where {O,T,M}
+  @assert 1 ≤ k ≤ length(locations) "number of neighbors must be smaller than number of locations"
+  @assert locations ⊆ 1:npoints(object) "invalid locations for spatial object"
+  kdtree = KDTree(coordinates(object, locations), ball.metric)
+  KBallSearcher{O,typeof(kdtree),T,M}(object, k, ball, locations, kdtree)
+end
+
+maxneighbors(searcher::KBallSearcher) = searcher.nmax
```
Comment (juliohm) [inline src/neighborsearch/kball_searcher.jl:36]: I think this is outdated code, right?

File: src/neighborsearch/kball_searcher.jl [hunk 12]
```diff
@@ -0,0 +1,68 @@
...
+  # state fields
+  kdtree::K
+
+end
+
+
+function KBallSearcher(object::O, k::Int, ball::BallNeighborhood{T,M}; locations=1:npoints(object)) where {O,T,M}
+  @assert 1 ≤ k ≤ length(locations) "number of neighbors must be smaller than number of locations"
+  @assert locations ⊆ 1:npoints(object) "invalid locations for spatial object"
+  kdtree = KDTree(coordinates(object, locations), ball.metric)
```
Comment (juliohm) [inline src/neighborsearch/kball_searcher.jl:32]: We try to avoid accessing fields of objects directly. In this case we have `metric(ball)` defined. Could you please use that instead?

File: src/neighborsearch/kball_searcher.jl [hunk 13]
```diff
@@ -0,0 +1,68 @@
...
+
+"""
+    KBallSearcher(object, k, ball; locations)
+
+A method for searching neighbors in spatial `object` inside `neighborhood`.
+Used when doing estimation combining k-nearest and radius.
+It searches first for the k-nearest by using the internal kdtree and then
+limiting the results within the `BallNeighborhood`
+"""
+struct KBallSearcher{O,K,T,M} <: AbstractBoundedNeighborSearcher
```
Comment (juliohm) [inline src/neighborsearch/kball_searcher.jl:15]: I think we can simplify the type parameters here. I've learned over time that we sometimes parameterize things too heavily given our C++ background. The idea is to parameterize with the most generic types possible so that the code works with types that we didn't imagine at first:

```julia
struct KBallSearcher{O,B,K} <: AbstractBoundedNeighborSearcher
  object::O
  k::Int
  ball::B
  locations::Vector{Int}
  
  # state fields
  kdtree::K
end
```

File: src/neighborsearch/kball_searcher.jl [hunk 14]
```diff
@@ -0,0 +1,68 @@
...
+  ball::BallNeighborhood{T,M}
+  locations::Vector{Int}
+
+  # state fields
+  kdtree::K
+
+end
+
+
+function KBallSearcher(object::O, k::Int, ball::BallNeighborhood{T,M}; locations=1:npoints(object)) where {O,T,M}
```
Comment (juliohm) [inline src/neighborsearch/kball_searcher.jl:29]: We can simplify this signature as well to something like:

```julia
function KBallSearcher(object::O, k::Int, ball::B; locations=1:npoints(object)) where {O,B}
  # ...
  KBallSearcher{O,B,typeof(kdtree)}(object, k, ball, locations, kdtree)
end
```

File: src/neighborsearch/kball_searcher.jl [hunk 15]
```diff
@@ -0,0 +1,68 @@
...
+end
+
+maxneighbors(searcher::KBallSearcher) = searcher.nmax
+
+function search!(neighbors::AbstractVector{Int}, xₒ::AbstractVector,
+                 searcher::KBallSearcher; mask=nothing)
+
+  k      = searcher.k
+  ball   = searcher.ball
+  radius = ball.radius
```
Comment (juliohm) [inline src/neighborsearch/kball_searcher.jl:43]: Please use `radius(ball)` instead

File: src/neighborsearch/kball_searcher.jl [hunk 16]
```diff
@@ -0,0 +1,68 @@
...
+
+  inds, dists = knn(searcher.kdtree, xₒ, k, true)
+  locs = view(searcher.locations, inds) #translate from relative indices (inds) to pdomain indices
```
Comment (juliohm) [inline src/neighborsearch/kball_searcher.jl:46]: Do you mind removing the comment? I think it helps somehow but it also can be inferred from the constructor easily. Also it contains this `pdomain` jargon people may not follow :relaxed:

File: src/neighborsearch/kball_searcher.jl [hunk 17]
```diff
@@ -0,0 +1,68 @@
...
+
+  if mask ≠ nothing
+    nneigh = 0
+    @inbounds for i in 1:length(inds)
```
Comment (juliohm) [inline src/neighborsearch/kball_searcher.jl:50]: Maybe replace `length(inds)` by `k`

File: src/neighborsearch/kball_searcher.jl [hunk 18]
```diff
@@ -0,0 +1,68 @@
...
+      #if mask is true, the locations is already
```
Comment (juliohm) [inline src/neighborsearch/kball_searcher.jl:51]: mind removing this comment as well?

File: src/neighborsearch/kball_searcher.jl [hunk 19]
```diff
@@ -0,0 +1,68 @@
...
+      if mask[locs[i]] && (dists[i] ≤ radius)
+        nneigh += 1
+        neighbors[nneigh] = locs[i]
+      end
+    end
+  else
+    nneigh = 0
+    @inbounds for i in 1:length(inds)
```
Comment (juliohm) [inline src/neighborsearch/kball_searcher.jl:59]: replace `length(inds)` by `k`

File: src/neighborsearch/kball_searcher.jl [hunk 20]
```diff
@@ -0,0 +1,68 @@
...
+      if dists[i] ≤ radius
+        nneigh += 1
+        neighbors[nneigh] = locs[i]
+      end
+    end
+  end
+
+  nneigh
+end
```
Comment (juliohm) [inline src/neighborsearch/kball_searcher.jl:68]: I never seen GitHub add this newline warning in a review, interesting. Do you think we should add a new line like in other files?

File: src/neighborsearch/kball_searcher.jl [hunk 21]
```diff
@@ -0,0 +1,68 @@
...
+                 searcher::KBallSearcher; mask=nothing)
+
+  k      = searcher.k
+  ball   = searcher.ball
+  radius = ball.radius
+
+  inds, dists = knn(searcher.kdtree, xₒ, k, true)
+  locs = view(searcher.locations, inds) #translate from relative indices (inds) to pdomain indices
+
+  if mask ≠ nothing
```
Comment (juliohm) [inline src/neighborsearch/kball_searcher.jl:48]: This `mask` business I've introduced some time ago needs to be refactored at some point in the future. The implementation feels clumsy because of this mask, yet I don't have any idea at the moment to replace this feature by something clever. Let me know if you have something in mind.

File: test/neighborsearch.jl [hunk 22]
```diff
@@ -0,0 +1,59 @@
+@testset "Neighborsearch" begin
+  @testset "KBallSearcher" begin
+    # KBallSearcher
+
```
Comment (juliohm) [inline test/neighborsearch.jl:4]: Can erase this comment as we have the testset already :+1:

File: test/neighborsearch.jl [hunk 23]
```diff
@@ -0,0 +1,59 @@
...
+    #create a regular grid to test
+    pdomain = RegularGrid(10,10)
```
Comment (juliohm) [inline test/neighborsearch.jl:6]: I would call this `D` or `dom` or something like it. The `pdomain` jargon was introduced in some solvers just to mean "problem domain". Here it doesn't feel very natural.

File: test/neighborsearch.jl [hunk 24]
```diff
@@ -0,0 +1,59 @@
...
+
+    #test 1: all points should be found
+    k = 10
+    ball = BallNeighborhood(100.0)
+    kballsearcher = KBallSearcher(pdomain, k, ball)
+
+    neighbors = neighbors = Vector{Int}(undef, k)
```
Comment (juliohm) [inline test/neighborsearch.jl:13]: Could you please double check this neighbors = neighbors here?

File: test/neighborsearch.jl [hunk 25]
```diff
@@ -0,0 +1,59 @@
...
+    xₒ = [5.0,5.0]
+    nneigh = search!(neighbors,xₒ,kballsearcher; mask=nothing)
```
Comment (juliohm) [inline test/neighborsearch.jl:15]: you can omit the mask argument since it defaults to nothing, it will make code cleaner and if we find a better strategy than this mask in the future the test will still work.

Comment (juliohm) [reply test/neighborsearch.jl:15]: also instead of testing the low-level `search!` here you could be testing the allocating version `search` that has a default implementation calling `search!`. You can just do `neighbors = search(xo, kballsearcher)`

File: test/neighborsearch.jl [hunk 26]
```diff
@@ -0,0 +1,59 @@
...
+    @test nneigh == 10
+
+    #test 2: radius of 1.0 m 5 points should be found
+    k = 10
+    ball = BallNeighborhood(1.0)
+    kballsearcher = KBallSearcher(pdomain, k, ball)
+
+    neighbors = Vector{Int}(undef, k)
+    xₒ = [5.0,5.0]
+    nneigh = search!(neighbors,xₒ,kballsearcher; mask=nothing)
```
Comment (juliohm) [inline test/neighborsearch.jl:26]: sample applies to here, you could do `neighbors = search(xo, kballsearcher)` and then `nneigh = length(neighbors)`

File: src/neighborsearch/kball_searcher.jl (hunk 12 above)
Comment (exepulveda) [reply src/neighborsearch/kball_searcher.jl:32]: Sure. Could we explain me the reasons?

Comment (juliohm) [reply src/neighborsearch/kball_searcher.jl:32]: It is because the object `ball` has a publicly defined API that we try to use. At some point in the future if the field name changes to something else, we will still be able to call `metric(ball)`. We only access fields directly when we are implementing a method for a function that is related to the struct we are defining the behavior. For example, in this case we are implementing the `searcher` so we can access the fields directly. However, the field `ball` of the searcher has its behavior defined elsewhere in the source and so we should only rely on the public api. Makes sense?

**Comment by juliohm** [comment]
@exepulveda I remembered this issue today, and I think we may be overcomplicating the solution? We already have two searchers: (1) k-nearest neighbors and (2) ball neighborhood. The neighborhood searcher returns all the locations inside the ball, no matter the number. The problem in the Kriging solver is that we are fitting this neighborhood searcher to the entire `pdomain` as opposed to the `pdata` locations. If we construct the searcher with `pdata` we solve the issue almost completely. The only missing part is the treatment of `minneighbors` and `maxneighbors` parameters. We could follow your suggestion and leave the cases with `n < minneighbors` untouched, and randomly select `maxneighbors` from the pool of `n` without replacement to make sure that we never exceed this value. What do you think?

In summary, the fixes that we could implement quickly in Kriging are:

- Fit the existing neighborhood searcher to `pdata`
- In the Kriging loop, get all the `n` neighbors
- If `n < minneighbors`, then place `missing` at location
- Otherwise sample `min(n, maxneighbors)` locations without replacement

If we do this, we have a faster Kriging already with neighborhoods that we could iterate more later with more improvements. This would also avoid introducing a new searcher like in this PR.

**Comment by exepulveda** [comment]
That is partially correct. Remember that (1) and (2) can be combined and it will work faster (that was the original idea with the new searcher). I agree with indexing pdata instead will fix the issue in 95% roughly. If you don't mind let me try to do it. I can at the same time learn the inside of GeoStatsBase.

**Comment by juliohm** [comment]
I added a few more fixes to the implementation, and by accident also added other commits from the master branch. Sorry for polluting the PR, I will merge and reconcile in the master branch. :+1:
